### PR TITLE
12230: Add data validation to GIDS on VACautionFlag.csv file upload

### PR DESCRIPTION
**Pull request by elder-joseph**
## Description
As an EDU team member, I need validation to made aware of potential data issues when uploading the CSV file so that I will know if there are any issues with the data in the file that need to be corrected before publishing a new version.

[Originating Issue](https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/va.gov-team/12230)

## Testing done
CI tests pass locally.

## Screenshots
N/A

## Acceptance criteria
- [x] The following validation message is displayed on upload of the VACautionFlag.csv file: "Row X: Both settlement title and settlement description are required for the settlement caution flag to be displayed." when one of the following is blank:
a. "settlement_title"
b. "settlement_description"
- [x] The following validation message is displayed on upload of the VACautionFlag.csv file: "Row X: The row has settlement data, but does not have a Title and Description." when the following occurs:
a. The file contains "settlement_date" OR "settlement_link" data
b. The file does not contain the "settlement_title" AND "settlement_description" data
- [x] If there are no settlement values in a row, and the row has "School_Closing_Date" or "Sec_702" data, no validation message is displayed.

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/models/va_caution_flag.rb [hunk 1]
```diff
@@ -16,6 +16,8 @@ class VaCautionFlag < ApplicationRecord
 
   validates :facility_code, presence: true
   validate :validate_date_fields
+  validate :validate_settlement_data
```
Comment (zurbergram) [inline app/models/va_caution_flag.rb:19]: Move all of this into validator class, see https://github.com/department-of-veterans-affairs/gibct-data-service/blob/master/app/models/edu_program.rb for how to setup

File: app/models/va_caution_flag.rb [hunk 2]
```diff
@@ -16,6 +16,8 @@ class VaCautionFlag < ApplicationRecord
...
+  validate :validate_school_closing_sec702
```
Comment (zurbergram) [inline app/models/va_caution_flag.rb:20]: Move all of this into validator class, see https://github.com/department-of-veterans-affairs/gibct-data-service/blob/master/app/models/edu_program.rb for how to setup

File: app/models/va_caution_flag.rb [hunk 3]
```diff
@@ -31,4 +33,20 @@ def validate_date_fields
...
+    if (!settlement_title.nil? || !settlement_description.nil? || !settlement_link.nil? || !settlement_date.nil?) && (settlement_title.nil? || settlement_description.nil?)
+      if settlement_link.nil? && settlement_date.nil?
+        errors.add(:base, 'Both settlement title and settlement description are required fo the settlement caution flag to be displayed.')
+      else
+        errors.add(:base, 'The row has settlement data, but does not have a Title and Description.')
+      end
+    end
+  end
+
+  def validate_school_closing_sec702
```
Comment (zurbergram) [inline app/models/va_caution_flag.rb:47]: This case in the function is not covered by the story

Comment (zurbergram) [reply app/models/va_caution_flag.rb:47]: Method name does not match method action

File: app/models/va_caution_flag.rb [hunk 4]
```diff
@@ -31,4 +33,20 @@ def validate_date_fields
       errors.add(:school_closing_date, 'must be mm/dd/yy')
     end
   end
+
+  def validate_settlement_data
+    if (!settlement_title.nil? || !settlement_description.nil? || !settlement_link.nil? || !settlement_date.nil?) && (settlement_title.nil? || settlement_description.nil?)
```
Comment (zurbergram) [inline app/models/va_caution_flag.rb:38]:
```suggestion
    if (!settlement_title.nil? || !settlement_description.nil? || !settlement_link.nil? || !settlement_date.nil?)
```

File: app/models/va_caution_flag.rb [hunk 5]
```diff
@@ -31,4 +33,20 @@ def validate_date_fields
...
+      if settlement_link.nil? && settlement_date.nil?
```
Comment (zurbergram) [inline app/models/va_caution_flag.rb:39]:
```suggestion
      if settlement_title.nil? && settlement_description.nil?
```

File: app/models/va_caution_flag.rb [hunk 6]
```diff
@@ -31,4 +33,20 @@ def validate_date_fields
...
+        errors.add(:base, 'Both settlement title and settlement description are required fo the settlement caution flag to be displayed.')
+      else
```
Comment (zurbergram) [inline app/models/va_caution_flag.rb:41]:
```suggestion
      elsif settlement_title.nil? ||settlement_description.nil?
```

File: app/models/va_caution_flag.rb [hunk 7]
```diff
@@ -31,4 +33,20 @@ def validate_date_fields
       errors.add(:school_closing_date, 'must be mm/dd/yy')
     end
   end
+
+  def validate_settlement_data
```
Comment (zurbergram) [inline app/models/va_caution_flag.rb:37]: This logic seems more complex then needed, could be written in a cleaner way

File: app/models/va_caution_flag.rb (hunk 3 above)
Comment (zurbergram) [reply app/models/va_caution_flag.rb:47]: Method name matches AC but method action doesn't

**Review by zurbergram — CHANGES_REQUESTED** [review]
Move validator code into a validator and address issues with complexity and AC

File: app/models/va_caution_flag.rb (hunk 2 above)
Comment (elder-joseph) [reply app/models/va_caution_flag.rb:20]: Will pull that out into the validator class

File: app/validators/va_caution_flag_validator.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class VaCautionFlagValidator < ActiveModel::Validator
+  def validate(record)
+    if !record.settlement_title.nil? || !record.settlement_description.nil? || !record.settlement_link.nil? || !record.settlement_date.nil?
+      if record.settlement_title.nil? && record.settlement_description.nil?
+        record.errors[:base] << 'Both record.settlement title and record.settlement description are required fo the record.settlement caution flag to be displayed.'
```
Comment (zurbergram) [inline app/validators/va_caution_flag_validator.rb:7]: `record.` ?

Comment (zurbergram) [inline app/validators/va_caution_flag_validator.rb:7]: `record.` in the message?

**Review by zurbergram — APPROVED** [review]
Error message has a `record.` in it and could use `.present?` instead of `!~.nil?`

**Review by sonntag-adam — APPROVED** [review]

**Review by thilton-oddball — APPROVED** [review]
Though it technically works, you may want to simplify the validate method. Otherwise looks good.

**Pull request merged**